Signal/src/util/DateUtil.m

```diff
@@ -187,19 +206,132 @@ + (NSString *)formatDateShort:(NSDate *)date
+        = MAX(0, [[calendar components:NSCalendarUnitMinute fromDate:date toDate:nowDate options:0] minute]);
+    NSInteger hoursDiff
+        = MAX(0, [[calendar components:NSCalendarUnitHour fromDate:date toDate:nowDate options:0] hour]);
+
+    NSString *result;
+    if (yearsDiff > 0) {
+        // "Long date" + locale-specific "short" time format.
+        NSString *dayOfWeek = [self.otherYearMessageFormatter stringFromDate:date];
+        NSString *formattedTime = [[self timeFormatter] stringFromDate:date];
+        result = [[dayOfWeek rtlSafeAppend:@" "] rtlSafeAppend:formattedTime];
+    } else if (daysDiff >= 7) {
+        // "Short date" + locale-specific "short" time format.
+        NSString *dayOfWeek = [self.thisYearMessageFormatter stringFromDate:date];
+        NSString *formattedTime = [[self timeFormatter] stringFromDate:date];
+        result = [[dayOfWeek rtlSafeAppend:@" "] rtlSafeAppend:formattedTime];
+    } else if (daysDiff > 0) {
+        // "Day of week" + locale-specific "short" time format.
+        NSString *dayOfWeek = [self.thisWeekMessageFormatter stringFromDate:date];
+        NSString *formattedTime = [[self timeFormatter] stringFromDate:date];
+        result = [[dayOfWeek rtlSafeAppend:@" "] rtlSafeAppend:formattedTime];
+    } else if (minutesDiff < 1) {
+        result = NSLocalizedString(@"DATE_NOW", @"The present; the current time.");
+    } else if (hoursDiff < 1) {
+        NSString *minutesString = [OWSFormat formatInt:(int)minutesDiff];
+        result = [NSString stringWithFormat:NSLocalizedString(@"DATE_MINUTES_AGO_FORMAT",
+                                                @"Format string for a relative time, expressed as a certain number of "
+                                                @"minutes in the past. Embeds {{The number of minutes}}."),
+                           minutesString];
+    } else {
+        NSString *hoursString = [OWSFormat formatInt:(int)hoursDiff];
+        result = [NSString stringWithFormat:NSLocalizedString(@"DATE_HOURS_AGO_FORMAT",
+                                                @"Format string for a relative time, expressed as a certain number of "
+                                                @"hours in the past. Embeds {{The number of hours}}."),
+                           hoursString];
+    }
+    return result.uppercaseString;
+}
+
++ (BOOL)isTimestampFromLastHour:(uint64_t)timestamp
+{
+    NSDate *date = [NSDate ows_dateWithMillisecondsSince1970:timestamp];
+    uint64_t nowTimestamp = [NSDate ows_millisecondTimeStamp];
+    NSDate *nowDate = [NSDate ows_dateWithMillisecondsSince1970:nowTimestamp];
+
+    NSCalendar *calendar = [NSCalendar currentCalendar];
+
+    NSInteger hoursDiff
+        = MAX(0, [[calendar components:NSCalendarUnitHour fromDate:date toDate:nowDate options:0] hour]);
+    return hoursDiff < 1;
+}
+
++ (NSString *)exemplaryNowTimeFormat
+{
+    return NSLocalizedString(@"DATE_NOW", @"The present; the current time.").uppercaseString;
```

Comment:
This probably deserves a comment. e.g. why "Now" instead of "June 3, 2017"

Reply:
I've added a comment in the header.